rust_src/src/fns.rs

```diff
@@ -375,4 +383,61 @@ pub fn load_average(use_floats: bool) -> Vec<LispNumber> {
         .collect()
 }
 
+/// Ask user a yes-or-no question.
+///
+/// Return t if answer is yes, and nil if the answer is no.  PROMPT is
+/// the string to display to ask the question.  It should end in a
+/// space; `yes-or-no-p' adds \"(yes or no) \" to it.
+///
+/// The user must confirm the answer with RET, and can edit it until
+/// it has been confirmed.
+///
+/// If dialog boxes are supported, a dialog box will be used if
+/// `last-nonmenu-event' is nil, and `use-dialog-box' is non-nil.
+#[lisp_fn]
+pub fn yes_or_no_p(prompt: LispStringRef) -> bool {
+    let use_popup = unsafe {
+        (globals.last_nonmenu_event.is_nil() || globals.last_nonmenu_event.is_cons())
+            && globals.use_dialog_box
+            && globals.last_input_event.is_not_nil()
+    };
+
+    if use_popup {
+        unsafe { redisplay_preserve_echo_area(4) };
+        let pane: LispObject = vec![("Yes", Qt).into(), ("No", Qnil).into()].into();
+        let menu: LispObject = (prompt, pane).into();
+        return unsafe { Fx_popup_dialog(Qt, menu, Qnil) }.into();
+    }
+
+    let yes_or_no: LispObject = "(yes or no) ".into();
+    let prompt = concat(&mut vec![prompt.into(), yes_or_no]).into();
+
+    loop {
+        let ans: LispStringRef = downcase(read_from_minibuffer(
+            prompt,
+            Qnil,
+            Qnil,
+            false,
+            Qyes_or_no_p_history,
+            Qnil,
+            false,
+        ))
+        .into();
+
+        if ans.as_slice() == "yes".as_bytes() {
```

Comment:
I added `slice_patterns` to the list of features recently. Can this be rewritten as a match?